# Homework 2
### Jake Pitkin
**CS 6190: Probabilistic Modeling - Spring 2018**<br>
**February 19 2018**

**1. Create the following Bayesian network to analyze the survey results. You will want to use the provided function createCPTfromData.**

<img src="img/network.png" width="855" height="611" align="center"/>

First we will create the following Bayesian network using the Behavioral Risk Factor Surveillance System
(BRFSS) survey data. The source code for the Bayesian network is located in `bayesian_network.py`.

In [4]:
from bayesian_network import *

# Read in the dataset
data = pd.read_csv("RiskFactors.csv")

# Create the Bayesian network from the dataset
income = createCPTfromData(data, ["income"])
bmi = createCPTfromData(data, ["bmi", "income", "exercise"])
exercise = createCPTfromData(data, ["exercise", "income"])
smoke = createCPTfromData(data, ["smoke", "income"])
bp = createCPTfromData(data, ["bp", "income", "exercise", "smoke"])
cholesterol = createCPTfromData(data, ["cholesterol", "income", "smoke", "exercise"])
diabetes = createCPTfromData(data, ["diabetes", "bmi"])
stroke = createCPTfromData(data, ["stroke", "bmi", "bp", "cholesterol"])
attack = createCPTfromData(data, ["attack", "bmi", "bp", "cholesterol"])
angina = createCPTfromData(data, ["angina", "bmi", "bp", "cholesterol"])
network = [income, bmi, exercise, smoke, bp, cholesterol, diabetes, stroke, attack, angina]
variables = ["income", "bmi", "exercise", "smoke", "bp", "cholesterol", "diabetes", "stroke", "attack", "angina"]

**1a. What is the size (in terms of the number of probabilities needed) of this network?**

The size of the network can be determined by considering the number of rows in each factor of the network.

In [16]:
rows = 0
for factor in network:
    rows += factor.shape[0]
print(rows)

504


The size of this network is 504 probabilities.

**1b. What is the total number of probabilities needed to store the full joint distribution?**

For the entire joint distribution, we take the product of the number of values each random variable can take:

$$\prod_{i=1}^{10} \ n(domain_i) = 8 * 2 * 2 * 4 * 4 * 2 * 2 * 2 * 2 * 4 = 32,768$$

The size of the full join distribution is 32,768 probabilities.

**2. For each of the four health outcomes (diabetes, stroke, heart attack, angina), answer the following by querying your network (using your infer function):**

**2a. What is the probabilitiy of the outcome if I have bad habits (smoke and don't exercise)?**

In [17]:
diabetes_marg = [x for x in variables if x != "diabetes"]
stroke_marg = [x for x in variables if x != "stroke"]
heart_attack_marg = [x for x in variables if x != "attack"]
angina_marg = [x for x in variables if x != "angina"]

diabetes_bad_habits = infer(network, diabetes_marg, ["smoke", "exercise"], [1, 2])
stroke_bad_habits = infer(network, stroke_marg, ["smoke", "exercise"], [1, 2])
heart_attack_bad_habits = infer(network, heart_attack_marg, ["smoke", "exercise"], [1, 2])
angina_bad_habits = infer(network, angina_marg, ["smoke", "exercise"], [1, 2])

print(diabetes_bad_habits)
print(stroke_bad_habits)
print(heart_attack_bad_habits)
print(angina_bad_habits)

      probs  diabetes
0  0.159330         1
1  0.007881         2
2  0.812632         3
3  0.020156         4
      probs  stroke
0  0.050127       1
1  0.949873       2
      probs  attack
0  0.072415       1
1  0.927585       2
      probs  angina
0  0.077756       1
1  0.922244       2


<table style="width: 80%, text-align: center">
  <caption style="font-size: 14px; color: black; text-align: center">Health Risks With Bad Habits</caption>
  <tr style="border: 1px solid black">
    <th style="border: 1px solid black"></th>
    <th style="border: 1px solid black">Diabetes</th>
    <th style="border: 1px solid black">Stroke</th>
    <th style="border: 1px solid black">Heart Attack</th>
    <th style="border: 1px solid black">Angina</th>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Yes</td>
    <td style="border: 1px solid black">0.159330</td>
    <td style="border: 1px solid black">0.050127</td>
    <td style="border: 1px solid black">0.072515</td>
    <td style="border: 1px solid black">0.077756</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">No</td>
    <td style="border: 1px solid black">0.812632</td>
    <td style="border: 1px solid black">0.949873</td>
    <td style="border: 1px solid black">0.927585</td>
    <td style="border: 1px solid black">0.922244</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Only During Pregnancy</td>
    <td style="border: 1px solid black">0.007881</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Pre-diabetic</td>
    <td style="border: 1px solid black">0.020156</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
</table>

**2a. How about if I have good habits (don't smoke and do exercise)**

In [29]:
diabetes_good_habits = infer(network, diabetes_marg, ["smoke", "exercise"], [2, 1])
stroke_good_habits = infer(network, stroke_marg, ["smoke", "exercise"], [2, 1])
heart_attack_good_habits = infer(network, heart_attack_marg, ["smoke", "exercise"], [2, 1])
angina_good_habits = infer(network, angina_marg, ["smoke", "exercise"], [2, 1])

print(diabetes_good_habits)
print(stroke_good_habits)
print(heart_attack_good_habits)
print(angina_good_habits)

      probs  diabetes
0  0.135062         1
1  0.007674         2
2  0.839392         3
3  0.017873         4
      probs  stroke
0  0.036808       1
1  0.963192       2
      probs  attack
0  0.051027       1
1  0.948973       2
      probs  angina
0  0.052319       1
1  0.947681       2


<table style="width: 80%, text-align: center">
  <caption style="font-size: 14px; color: black; text-align: center">Health Risks With Good Habits</caption>
  <tr style="border: 1px solid black">
    <th style="border: 1px solid black"></th>
    <th style="border: 1px solid black">Diabetes</th>
    <th style="border: 1px solid black">Stroke</th>
    <th style="border: 1px solid black">Heart Attack</th>
    <th style="border: 1px solid black">Angina</th>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Yes</td>
    <td style="border: 1px solid black">0.135062</td>
    <td style="border: 1px solid black">0.036808</td>
    <td style="border: 1px solid black">0.051027</td>
    <td style="border: 1px solid black">0.052319</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">No</td>
    <td style="border: 1px solid black">0.839392</td>
    <td style="border: 1px solid black">0.963192</td>
    <td style="border: 1px solid black">0.948973</td>
    <td style="border: 1px solid black">0.947681</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Only During Pregnancy</td>
    <td style="border: 1px solid black">0.007674</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Pre-diabetic</td>
    <td style="border: 1px solid black">0.017873</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
</table>

**2b. What is the probability of the outcome if I have poor health (high blood pressure, high cholesterol, and overweight)?**

In [30]:
diabetes_poor_health = infer(network, diabetes_marg, ["bp", "cholesterol", "bmi"], [1, 1, 3])
stroke_poor_health = infer(network, stroke_marg, ["bp", "cholesterol", "bmi"], [1, 1, 3])
heart_attack_poor_health = infer(network, heart_attack_marg, ["bp", "cholesterol", "bmi"], [1, 1, 3])
angina_poor_health = infer(network, angina_marg, ["bp", "cholesterol", "bmi"], [1, 1, 3])

print(diabetes_poor_health)
print(stroke_poor_health)
print(heart_attack_poor_health)
print(angina_poor_health)

      probs  diabetes
0  0.122279         1
1  0.006718         2
2  0.854003         3
3  0.017000         4
      probs  stroke
0  0.083975       1
1  0.916025       2
     probs  attack
0  0.13433       1
1  0.86567       2
      probs  angina
0  0.153185       1
1  0.846815       2


<table style="width: 80%, text-align: center">
  <caption style="font-size: 14px; color: black; text-align: center">Health Risks With Poor Health</caption>
  <tr style="border: 1px solid black">
    <th style="border: 1px solid black"></th>
    <th style="border: 1px solid black">Diabetes</th>
    <th style="border: 1px solid black">Stroke</th>
    <th style="border: 1px solid black">Heart Attack</th>
    <th style="border: 1px solid black">Angina</th>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Yes</td>
    <td style="border: 1px solid black">0.122279</td>
    <td style="border: 1px solid black">0.083975</td>
    <td style="border: 1px solid black">0.13433</td>
    <td style="border: 1px solid black">0.153185</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">No</td>
    <td style="border: 1px solid black">0.854003</td>
    <td style="border: 1px solid black">0.916025</td>
    <td style="border: 1px solid black">0.86567</td>
    <td style="border: 1px solid black">0.846815</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Only During Pregnancy</td>
    <td style="border: 1px solid black">0.006718</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Pre-diabetic</td>
    <td style="border: 1px solid black">0.017</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
</table>

**2b. How about if I have good health (low blood pressure, low cholesterol, and normal weight)?**

In [33]:
diabetes_good_health = infer(network, diabetes_marg, ["bp", "cholesterol", "bmi"], [3, 2, 2])
stroke_good_health = infer(network, stroke_marg, ["bp", "cholesterol", "bmi"], [3, 2, 2])
heart_attack_good_health = infer(network, heart_attack_marg, ["bp", "cholesterol", "bmi"], [3, 2, 2])
angina_good_health = infer(network, angina_marg, ["bp", "cholesterol", "bmi"], [3, 2, 2])

print(diabetes_good_health)
print(stroke_good_health)
print(heart_attack_good_health)
print(angina_good_health)

      probs  diabetes
0  0.061634         1
1  0.007800         2
2  0.919897         3
3  0.010668         4
     probs  stroke
0  0.01387       1
1  0.98613       2
      probs  attack
0  0.015888       1
1  0.984112       2
      probs  angina
0  0.012839       1
1  0.987161       2


<table style="width: 80%, text-align: center">
  <caption style="font-size: 14px; color: black; text-align: center">Health Risks With Good Health</caption>
  <tr style="border: 1px solid black">
    <th style="border: 1px solid black"></th>
    <th style="border: 1px solid black">Diabetes</th>
    <th style="border: 1px solid black">Stroke</th>
    <th style="border: 1px solid black">Heart Attack</th>
    <th style="border: 1px solid black">Angina</th>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Yes</td>
    <td style="border: 1px solid black">0.061634</td>
    <td style="border: 1px solid black">0.01387</td>
    <td style="border: 1px solid black">0.01588</td>
    <td style="border: 1px solid black">0.012839</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">No</td>
    <td style="border: 1px solid black">0.919897</td>
    <td style="border: 1px solid black">0.98613</td>
    <td style="border: 1px solid black">0.984112</td>
    <td style="border: 1px solid black">0.987161</td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Only During Pregnancy</td>
    <td style="border: 1px solid black">0.0078</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
  <tr style="border: 1px solid black">
    <td style="border: 1px solid black; font-weight: bold">Pre-diabetic</td>
    <td style="border: 1px solid black">0.010668</td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
    <td style="border: 1px solid black"></td>
  </tr>
</table>